haystack_experimental/components/splitters/hierarchical_doc_splitter.py

```diff
@@ -0,0 +1,141 @@
+    >> Document(id=52c7e9fc53ae9aa734cc15d8624ae63468b423a7032077ee4cdcf524569274d3, content: 'a ', meta: {'block_size': 2, 'parent_id': '5ff4a36b5371580ac5f2ea21a5690dcc18802c7e5e187d57c5e2d312eee22dfd', 'children_ids': [], 'level': 2, 'source_id': '5ff4a36b5371580ac5f2ea21a5690dcc18802c7e5e187d57c5e2d312eee22dfd', 'page_number': 1, 'split_id': 1, 'split_idx_start': 8}),
+    >> Document(id=39d299629a35051fc9ebb62a0594626546d6ec1b1de7cfcdfb03be58b769478e, content: 'simple test ', meta: {'block_size': 2, 'parent_id': '8dc5707ebe647dedab97db15d0fc82a5e551bbe54a9fb82f79b68b3e3046a3a2', 'children_ids': [], 'level': 2, 'source_id': '8dc5707ebe647dedab97db15d0fc82a5e551bbe54a9fb82f79b68b3e3046a3a2', 'page_number': 1, 'split_id': 0, 'split_idx_start': 0}),
+    >> Document(id=e23ceb261772f539830384097e4f6c513205019cf422378078ff66dd4870f91a, content: 'document', meta: {'block_size': 2, 'parent_id': '8dc5707ebe647dedab97db15d0fc82a5e551bbe54a9fb82f79b68b3e3046a3a2', 'children_ids': [], 'level': 2, 'source_id': '8dc5707ebe647dedab97db15d0fc82a5e551bbe54a9fb82f79b68b3e3046a3a2', 'page_number': 1, 'split_id': 1, 'split_idx_start': 12})]}
+    ```
+    """  # noqa: E501
+
+    def __init__(
+        self,
+        block_sizes: List[int],
+        split_overlap: int = 0,
+        split_by: Literal["word", "sentence", "page", "passage"] = "word",
+    ):
+        """
+        Initialize HierarchicalDocumentBuilder.
+
+        :param block_sizes: List of block sizes to split the document into. The blocks are split in descending order.
+        :param split_overlap: The number of overlapping units for each split.
+        :param split_by: The unit for splitting your documents.
+        """
+
+        if len(set(block_sizes)) != len(block_sizes):
+            raise ValueError("block_sizes must not contain duplicates")
+        self.block_sizes = sorted(set(block_sizes), reverse=True)
+        self.split_overlap = split_overlap
+        self.split_by = split_by
+
+    @component.output_types(documents=List[Document])
+    def run(self, documents: List[Document]):
+        """
+        Builds a hierarchical document structure for each document in a list of documents.
+
+        :param documents: List of Documents to split into hierarchical blocks.
+        :return: List of HierarchicalDocument
+        """
+        hierarchical_docs = []
+        for doc in documents:
+            hierarchical_docs.extend(self.build_hierarchy_from_doc(doc))
+        return {"documents": hierarchical_docs}
+
+    def _split_doc(self, doc: Document, block_size: int) -> List[Document]:
+        splitter = DocumentSplitter(split_length=block_size, split_overlap=self.split_overlap, split_by=self.split_by)
+        split_docs = splitter.run([doc])
+        return split_docs["documents"]
+
+    @staticmethod
+    def _add_meta_data(document: Document):
+        document.meta["__block_size"] = 0
+        document.meta["__parent_id"] = None
+        document.meta["__children_ids"] = []
```

Comment:
Remember https://github.com/deepset-ai/haystack/issues/8181: vector DBs like Chroma and Pinecone do not support lists as metadata types. IDK if we want to ensure compatibility...

Reply:
Chroma accounts for around 3% of all the downloads related with DBs, Pinecone even less - I don't think rewriting the code will have any significant impact such that those can also be used with the Auto-Merge Retriever. If the number grows we can then rewrite the code so that those are also supported.

I will nevertheless add a warning

Reply:
Chroma is still among the most widely used DBs with Haystack.
And I would prefer to introduce new features while supporting existing integrations as much as possible, if the effort required is not too big.

Let's also hear what @shadeMe thinks.

Reply:
Please refer to my comment up top.